week3/homework/src/functions.js

```diff
@@ -0,0 +1,196 @@
+const fs = require('fs');
+const data = require('./index');
+const path = './to-dos.json';
+const fileName = 'to-dos.json';
+const uuidv4 = require('uuid/v4');
+const defaultFileContent = { todoArray: [] };
+
+function isFileExist() {
+  try {
+    if (fs.existsSync(path)) {
+      return true;
+    } else {
+      fs.writeFileSync(fileName, JSON.stringify(defaultFileContent));
+      return false;
+    }
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function isInputValid(request) {
+  try {
+    if (request.headers['content-type'] === 'application/json') {
+      if (request.body.hasOwnProperty('todo')) {
+        if (request.body.todo.hasOwnProperty('description')) {
+          if (
+            request.body.todo.description !== undefined &&
+            request.body.todo.description !== null &&
+            request.body.todo.description !== ''
+          ) {
+            return true;
+          }
+        }
+      }
+    } else return false;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function help() {
+  try {
+    return JSON.parse(fs.readFileSync('help.json', 'utf8'));
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function list() {
+  try {
+    if (isFileExist()) {
+      return JSON.parse(fs.readFileSync(fileName, 'utf8'));
+    }
+    return defaultFileContent;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function create(request) {
+  try {
+    if (isFileExist()) {
+      if (isInputValid(request)) {
+        request.body.todo.id = uuidv4();
+        request.body.todo.done = false;
+        const obj = JSON.parse(fs.readFileSync(fileName, 'utf8'));
+        obj.todoArray.push(request.body);
+        fs.writeFileSync(fileName, JSON.stringify(obj));
+        return obj;
+      }
+      return { error: 'content-type is not application/json or todo text is not valid' };
+    }
+    return defaultFileContent;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function getToDoById(id) {
+  try {
+    if (isFileExist()) {
+      const obj = JSON.parse(fs.readFileSync(fileName, 'utf8'));
+      const result = obj.todoArray.filter(element => {
+        if (element.todo.id == id) {
+          return element;
```

Comment:
I first thought, that this is a mistake. It turns out to work properly, but can you explain why?

Reply:
What is the mistake that You thought? maybe I can explain it after that